src/unix/mmap.c

```diff
@@ -681,6 +681,56 @@ static boolean mmap_reserve_range(process p, range q)
     return true;
 }
 
+closure_function(0, 1, void, msync_vmap,
+                 rmnode, n)
+{
+    vmap vm = (vmap)n;
+    if ((vm->flags & VMAP_FLAG_SHARED) &&
+        (vm->flags & VMAP_FLAG_MMAP) &&
+        (vm->flags & VMAP_MMAP_TYPE_MASK) == VMAP_MMAP_TYPE_FILEBACKED) {
+        assert(vm->cache_node);
+        pagecache_node_scan_shared_pages(vm->cache_node, n->r);
+    }
+}
+
+closure_function(1, 1, void, msync_gap,
+                 boolean *, have_gap,
+                 range, r)
+{
+    *bound(have_gap) = true;
+}
+
+static sysreturn msync(void *addr, u64 length, int flags)
+{
+    thread_log(current, "%s: addr %p, length 0x%lx, flags %x", __func__,
+               addr, length, flags);
+
+    if ((flags & ~(MS_ASYNC | MS_SYNC | MS_INVALIDATE)) ||
+        (flags & (MS_ASYNC | MS_SYNC)) == (MS_ASYNC | MS_SYNC))
+        return -EINVAL;
+
+    if (flags & MS_ASYNC)
+        return 0;               /* nop / default behavior */
+
+    /* TODO: Linux appears to only use this flag to test whether a map
+       is locked (and return -EBUSY if it is). We don't swap out
+       pages, so mlock is a stub for us. It's not clear if we will
+       later need to track locked status and test for this condition. */
+    if (flags & MS_INVALIDATE)
+        return 0;
```

Comment:
When Linux returns 0, it means the sync has been done, while if the sync is not being done (e.g. because of memory locks) an error is returned. Here we are returning 0 without actually doing the sync, which seems inconsistent, we should either ignore the MS_INVALIDATE flag, or return error (EBUSY) when the flag is specified. My preference would be to return EBUSY, since as you noted all of our mappings currently are resident in memory.

Reply:
> When Linux returns 0, it means the sync has been done, while if the sync is not being done (e.g. because of memory locks) an error is returned.

This is true if MS_SYNC is specified (not MS_ASYNC). The error here was in thinking that MS_INVALIDATE was mutually exclusive with the other flags. But the msync man page says for EBUSY: "MS_INVALIDATE was specified in flags, and a memory lock exists for the specified address range." File locking isn't applicable for us ... so the question is whether to default to returning 0 or treat all files as locked and return -EBUSY as you suggest. I'm going to assume the latter is the safest option for now. (Updated, and also added a default to ASYNC if neither sync flag is specified to match Linux behavior.)

There seems to be quite a bit of confusion as to the purpose of this flag. I wasn't able to find a use case for it from a cursory search. And Linux does not invalidate cache pages when this flag is specified; it only returns -EBUSY for locked pages.

https://groups.google.com/g/comp.unix.programmer/c/reGmzSHKBHE?pli=1
https://stackoverflow.com/questions/38689221/historical-current-use-of-ms-invalidate